elasticapm/contrib/starlette/__init__.py

```diff
@@ -0,0 +1,190 @@
+        >>> 'DEBUG': True,
+        >>> 'SERVER_URL': 'http://localhost:8200',
+        >>> 'CAPTURE_HEADERS': True,
+        >>> 'CAPTURE_BODY': 'all'
+    >>> })
+
+    >>> app.add_middleware(ElasticAPM, client=elasticapm)
+
+    Pass an arbitrary APP_NAME and SECRET_TOKEN::
+
+    >>> elasticapm = ElasticAPM(app, service_name='myapp', secret_token='asdasdasd')
+
+    Pass an explicit client::
+
+    >>> elasticapm = ElasticAPM(app, client=client)
+
+    Automatically configure logging::
+
+    >>> elasticapm = ElasticAPM(app, logging=True)
+
+    Capture an exception::
+
+    >>> try:
+    >>>     1 / 0
+    >>> except ZeroDivisionError:
+    >>>     elasticapm.capture_exception()
+
+    Capture a message::
+
+    >>> elasticapm.capture_message('hello, world!')
+    """
+
+    def __init__(self, app: ASGIApp, client: Client):
+        """
+
+        Args:
+            app (ASGIApp): Starlette app
+            client (Client): ElasticAPM Client
+        """
+        self.client = client
+
+        super().__init__(app)
+
+    async def dispatch(self, request: Request, call_next: RequestResponseEndpoint) -> Response:
+        """Processes the whole request APM capturing.
+
+        Args:
+            request (Request)
+            call_next (RequestResponseEndpoint): Next request process in Starlette.
+
+        Returns:
+            Response
+        """
+        await self._request_started(request)
+
+        try:
+            response = await call_next(request)
+        except Exception:
+            await self.capture_exception()
+            elasticapm.set_transaction_result("HTTP 5xx", override=False)
+            elasticapm.set_context({"status_code": 500}, "response")
+
+            raise
+        else:
+            await self._request_finished(response)
+        finally:
+            self.client.end_transaction()
+
+        return response
+
+    async def capture_exception(self, *args, **kwargs):
+        """Captures your exception.
+
+        Args:
+            *args:
+            **kwargs:
+        """
+        self.client.capture_exception(*args, **kwargs)
+
+    async def capture_message(self, *args, **kwargs):
+        """Captures your message.
+
+        Args:
+            *args: Whatever
+            **kwargs: Whatever
+        """
+        self.client.capture_message(*args, **kwargs)
+
+    async def _request_started(self, request: Request):
+        """Captures the begin of the request processing to APM.
+
+        Args:
+            request (Request)
+        """
+        trace_parent = TraceParent.from_headers(dict(request.headers))
+        self.client.begin_transaction("request", trace_parent=trace_parent)
+
+        request_data = await get_data_from_request(
+            request,
+            capture_body=self.client.config.capture_body in ("transactions", "all"),
+            capture_headers=self.client.config.capture_headers,
+        )
+
+        elasticapm.set_context(request_data, "request")
```

Comment:
We like to wrap these types of processing functions in a lambda, which `set_context()` only calls if the transaction is actually sampled. This way for unsampled transactions we can skip a lot of work, which is particularly relevant if we're capturing streaming bodies or the like.

Here's an example: https://github.com/elastic/apm-agent-python/blob/721601eec8abbdcbabf4d2567386012fdf9ba1bd/elasticapm/contrib/aiohttp/middleware.py#L67-L74

Reply:
I noticed that but `get_data_from_request()` is a coroutine in my case.
Updated..